### DISCUSSION TITLE
Hung Resources after upgrade

### DISCUSSION THREAD

**Discussion by arcarmona** [emqx/emqx#12511] [Q&A] [answered]
Hello,

I had an EMQx (emqx community edition 5.4.1) cluster set up on GKE using emqx operator with 3 nodes. Given 5.5.0 recently came out I decided to take the chance to test the blue green upgrade feature. After applying the following patch: `kubectl patch emqx emqx --type=merge -p '{"spec": {"image": "emqx:5.5.0"}}'`, the behavior at first was as expected, a new cluster was being generated which would end up taking over, and the old stateful set pods were being terminated, connections were being lost at first but after a brief time they reconnected to the new cluster and carried over the message load lost during this brief period of time and normal working conditions were met. 

What I can't seem to understand right now and the reason behind this post is the following:
- After deletion of the old stateful set pods, the stateful set on itself stayed running (with 0/0 pods it is) and so did the PVCs of the pods.
- In the dashboard it still showed the old stateful set nodes as part of the cluster, albeit as dead/stopped nodes (picture attached below)

I ended up manually deleting the old stateful set, however the issue with the dashboard persisted. My emqx.yaml can be found below the image. As for the deployment of the cluster I followed emqx-operator docs for deploying on GKE. Is there anyway to remove the stopped nodes from the dashboard and prevent this from happening next time I perform a patch? Should the old stateful set and PVCs stay up after patching?

Log of emqx-operator-controller-manager at the time of performing the patch (I forgot to save the logs of the old stateful set before deletion but could recreate and attach them if needed):

`
{"controller":"emqx", "controllerGroup":"apps.emqx.io", "controllerKind":"EMQX", "eMQX":{…}, "error":"failed to check if sts can be scale down: failed to get node info by API: failed to get API api/v5/nodes: failed to request API: Get "http://10.52.15.152:18083/api/v5/nodes/emqx@.emqx-headless.cloudservices.svc.cluster.local": dial tcp 10.52.15.152:18083: connect: connection refused", "errorVerbose":"Get "http://10.52.15.152:18083/api/v5/nodes/emqx@.emqx-headless.cloudservices.svc.cluster.local": dial tcp 10.52.15.152:18083: connect: connection refused
failed to request API
...
`

![image](https://github.com/emqx/emqx/assets/152393447/d862a725-d6f4-45fe-ae38-97a6fb1569d2)

```
apiVersion: apps.emqx.io/v2beta1
kind: EMQX
metadata:
  name: emqx
spec:
  image: emqx:5.4.1
  imagePullPolicy: Always
  updateStrategy:
    evacuationStrategy:
      connEvictRate: 1000
      sessEvictRate: 1000
      waitTakeover: 10
    initialDelaySeconds: 10
    type: Recreate
  config:
    data: |
      log.file {
        enable = true,
        formatter = "json"
      }
      listeners.tcp.default {
        proxy_protocol = true,
        max_connections = 10000,
        acceptors = 64,
      }
      listeners.ssl.default {
        enable = false
      }
      mqtt {
        max_topic_levels = 128,
        peer_cert_as_username = "cn",
        shared_subscription_strategy = "random",
        max_inflight = 2000
      }
      node {
        dist_buffer_size = 2097150,
        max_ports = 2097152
      }
      force_shutdown {
        enable = false,
      }
      retainer {
        enable = false,
      }
  coreTemplate:
    spec:
      resources:
        limits:
          cpu: "4"
          memory: 4Gi
        requests:
          cpu: "4"
          memory: 4Gi
      ports:
        - name: dashboard
          containerPort: 18083
      replicas: 3
      volumeClaimTemplates:
        storageClassName: standard
        resources:
          requests:
            storage: 10Gi
        accessModes:
        - ReadWriteOnce
  dashboardServiceTemplate:
    spec:
      type: ClusterIP
      ports:
        - name: dashboard
          port: 18083
          targetPort: 18083
  listenersServiceTemplate:
    spec:
      clusterIP: None # Headless
```

**Comment by id** [comment]
ping @Rory-Z

**Reply by arcarmona** [reply]
I ended up forcing the stopped nodes to leave the cluster but any idea on why this behavior happened?  @Rory-Z

**Reply by Rory-Z** [reply, marked as answer]
Hi, sorry for delay, I just finished my vacation.

For 0/0 sets and PVC, in the initial design, this was intended to roll back reserved resources, but we found it may be difficult to operate in actual deployment. We are also considering modifying it. And you can modify `.spec.revisionHistoryLimit` to change the number of 0/0 ReplicaSets, 0/0 StatefulSet and old PersistentVolumeClaim.

For EMQX stop node in cluster, because the default value of `cluster. autoclean` is 24h. @id Can we change is default value ? In EMQX 4, this default values is 5m

**Reply by id** [reply]
Sure, `cluster.autoclean` can be set in `spec.config.data` just as any other config.

**Reply by arcarmona** [reply]
Thank you for the support! I will mark as answered
